{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "import matplotlib as matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Data processing for Apparent Fitness Landscape'''\n",
    "\n",
    "def parse_from_txt(file):\n",
    "    '''This takes the aligned, merged file from terminal analysis. We first get rid of any genes that do not\n",
    "    start with ATG and end with TAA. Then we append the resulting genes to a list.'''\n",
    "    gene = \"\"\n",
    "    text_file = open(file, \"r\") #open and read the text file containing aligned library reads\n",
    "    lines = text_file.readlines()\n",
    "    text_file.close()\n",
    "    \n",
    "    record = [] #this is list of genes that will be counted later\n",
    "    for x in lines: #separate out each read to be analyzed individually\n",
    "        '''The start codon follow a tab, and the stop codon should preceed a tab'''\n",
    "        for i in range(100): #scan the first 100 letters for \\tATG\n",
    "            if x[i:i+4] == \"CATG\":\n",
    "                if x[i+391:i+394] == \"TAA\": #391 letters after the tab, there should be a stop codon and a tab\n",
    "                        gene = x[i+1:i+394] #only take records that agree with both of these conditions\n",
    "                        record.append(gene) #append to a list for further analysis\n",
    "    return record"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Data processing for Apparent Fitness Landscape'''\n",
    "\n",
    "def parse_from_txt_mini(file):\n",
    "    '''This takes the aligned, merged file from terminal analysis. We first get rid of any genes that do not\n",
    "    start with ATG and end with TAA. Then we append the resulting genes to a list.'''\n",
    "    gene = \"\"\n",
    "    text_file = open(file, \"r\") #open and read the text file containing aligned library reads\n",
    "    lines = text_file.readlines()\n",
    "    text_file.close()\n",
    "    \n",
    "    record = [] #this is list of genes that will be counted later\n",
    "    for x in lines: #separate out each read to be analyzed individually\n",
    "        '''The start codon follow a tab, and the stop codon should preceed a tab'''\n",
    "        for i in range(100): #scan the first 100 letters for \\tATG\n",
    "            if x[i:i+4] == \"CATG\":\n",
    "                if x[i+391:i+394] == \"TAA\": #391 letters after the tab, there should be a stop codon and a tab\n",
    "                    if x[i+112:i+115] == \"CCG\": #If there is the S37P mutation\n",
    "                        gene = x[i+1:i+394] #only take records that agree with both of these conditions\n",
    "                        record.append(gene) #append to a list for further analysis\n",
    "    return record"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Data processing for Apparent Fitness Landscape'''\n",
    "\n",
    "def count_heatmap(dna_lst):\n",
    "    '''This function takes the output of parse_from_txt and counts codons. Several conditions are in place:\n",
    "    First, we determine how many mutations there are per gene. We discard wildtype reads (or those with zero \n",
    "    mutations. Anything with one mutation is counted. For reads with two or more mutations, any 1-bp mutations\n",
    "    are discarded, while 2+ bp mutations are counted'''\n",
    "\n",
    "    ms2_gene = ('ATGGCTTCTAACTTTACTCAGTTCGTTCTCGTCGACAATGGCGGAAC'+\n",
    "    'TGGCGACGTGACTGTCGCCCCAAGCAACTTCGCTAACGGGGTCGCTGAATGGATCAGC'+\n",
    "    'TCTAACTCGCGTTCACAGGCTTACAAAGTAACCTGTAGCGTTCGTCAGAGCTCTGCGC'+\n",
    "    'AGAATCGCAAATACACCATCAAAGTCGAGGTGCCTAAAGTGGCAACCCAGACTGTTGG'+\n",
    "    'TGGTGTAGAGCTTCCTGTAGCCGCATGGCGTTCGTACTTAAATATGGAACTAACCATT'+\n",
    "    'CCAATTTTCGCTACGAATTCCGACTGCGAGCTTATTGTTAAGGCGATGCAAGGACTCC'+\n",
    "    'TAAAAGATGGAAACCCGATTCCCTCAGCAATCGCAGCAAACTCCGGCATCTACTAA')\n",
    "    \n",
    "    intensities = np.zeros((131, 64))\n",
    "\n",
    "    codon_to_num = {'TTT': 0, 'TTG': 1, 'TTA': 2, 'TTC': 3, \n",
    "                     'CTT': 4, 'CTG': 5, 'CTA': 6, 'CTC': 7,\n",
    "                     'ATT': 8, 'ATG': 9, 'ATA': 10, 'ATC': 11,\n",
    "                     'GTT': 12, 'GTG': 13, 'GTA': 14, 'GTC': 15,\n",
    "                     'TAT': 16, 'TAG': 17, 'TAA': 18, 'TAC': 19,\n",
    "                     'CAT': 20, 'CAG': 21, 'CAA': 22, 'CAC': 23,\n",
    "                     'AAT': 24, 'AAG': 25, 'AAA': 26, 'AAC': 27,\n",
    "                     'GAT': 28, 'GAG': 29, 'GAA': 30, 'GAC': 31,\n",
    "                     'TCT': 32, 'TCG': 33, 'TCA': 34, 'TCC': 35,\n",
    "                     'CCT': 36, 'CCG': 37, 'CCA': 38, 'CCC': 39,\n",
    "                     'ACT': 40, 'ACG': 41, 'ACA': 42, 'ACC': 43,\n",
    "                     'GCT': 44, 'GCG': 45, 'GCA': 46, 'GCC': 47,\n",
    "                     'TGT': 48, 'TGG': 49, 'TGA': 50, 'TGC': 51,\n",
    "                     'CGT': 52, 'CGG': 53, 'CGA': 54, 'CGC': 55,\n",
    "                     'AGT': 56, 'AGG': 57, 'AGA': 58, 'AGC': 59,\n",
    "                     'GGT': 60, 'GGG': 61, 'GGA': 62, 'GGC' :63} \n",
    "    \n",
    "    '''Each amino acid has a number associated with it, which is also its position in AAs.\n",
    "    i.e. A = 0; S = 1; etc. AA_code allows us to move back and forth between codon and AA #.\n",
    "    Lists checked on 6/27'''\n",
    "\n",
    "    AA_code = {44: 0, 45: 0, 46: 0, 47: 0, 32: 1, 33: 1, 34: 1, 35: 1, 56: 1, 59: 1, \n",
    "                 40: 2, 41: 2, 42: 2, 43: 2, 12: 3, 13: 3, 14: 3, 15: 3, 48: 4, 51: 4, \n",
    "                 29: 5, 30: 5, 28: 6, 31: 6, 25: 7, 26: 7, 52: 8, 53: 8, 54: 8, 55: 8, \n",
    "                 57: 8, 58: 8, 21: 9, 22: 9, 24: 10, 27: 10, 9: 11, 8: 12, 10: 12, 11: 12, \n",
    "                 1: 13, 2: 13, 4: 13, 5: 13, 6: 13, 7: 13, 20: 14, 23: 14, 0: 15, 3: 15, \n",
    "                 16: 16, 19: 16, 49: 17, 60: 18, 61: 18, 62: 18, 63: 18, 36: 19, 37: 19, \n",
    "                 38: 19, 39: 19, 17: 20, 18: 20, 50: 20}\n",
    "    \n",
    "    AAs = ['A', 'S', 'T', 'V', 'C', 'E', 'D', 'K', 'R', 'Q', 'N', 'M', 'I', 'L', \n",
    "           'H', 'F', 'Y', 'W', 'G', 'P', '*']\n",
    "    \n",
    "    for dna_read in dna_lst: #dna_read counts where you are in sequencing reads\n",
    "        '''first we count how many mutations there are in a read'''\n",
    "        number_of_mutations = 0 #number_of_mutations counts number of mutations in a dna_read (seq reads)\n",
    "        for codon_number in range(131): #codon_number is the residue number in MS2 CP\n",
    "            dna_base_num = codon_number * 3 #dna_base_num is dna base pair number\n",
    "            wildtype = ms2_gene[dna_base_num:dna_base_num+3]\n",
    "            experiment_codon = dna_read[dna_base_num:dna_base_num+3]\n",
    "            if wildtype != experiment_codon:\n",
    "                #counter increases if the codon is not wildtype\n",
    "                number_of_mutations += 1\n",
    "        if number_of_mutations > 1:\n",
    "            '''if number_of_mutations > 1, get rid of any mutations that are 1 bp different from wildtype'''\n",
    "            for second_loop_codon_number in range(131): #loop through read again\n",
    "                dna_base_num = second_loop_codon_number * 3 #get DNA base pair number\n",
    "                wildtype = ms2_gene[dna_base_num:dna_base_num+3] #slice out wildtype codon\n",
    "                experiment_codon = dna_read[dna_base_num:dna_base_num+3] #slice out experiment codon\n",
    "                read_level_mutation_counter = 0 #counter for the number of basepairs that are different from wildtype\n",
    "                if wildtype != experiment_codon: #if the two codons aren't the same, count # of differences\n",
    "                    for base in range(3):\n",
    "                        if experiment_codon[base] != wildtype[base]:\n",
    "                            read_level_mutation_counter += 1\n",
    "                    if read_level_mutation_counter == 1:\n",
    "                        pass #do not count the codon if it is a 1-bp mutation\n",
    "                    elif experiment_codon in codon_to_num: #otherwise, add it to the intesities array\n",
    "                        intensities[second_loop_codon_number, codon_to_num[experiment_codon]] += 1\n",
    "        if number_of_mutations == 0:\n",
    "            '''if wildtype, toss'''\n",
    "        if number_of_mutations == 1:\n",
    "            '''keep the read if there is one total mutation'''\n",
    "            for third_loop_codon_number in range(131):\n",
    "                dna_base_num = third_loop_codon_number * 3\n",
    "                wildtype = ms2_gene[dna_base_num:dna_base_num+3]\n",
    "                experiment_codon = dna_read[dna_base_num:dna_base_num+3]\n",
    "                if wildtype != experiment_codon: #find the mutation in the read\n",
    "                    intensities[third_loop_codon_number, codon_to_num[experiment_codon]] += 1\n",
    "    return intensities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Data processing for Apparent Fitness Landscape'''\n",
    "\n",
    "def count_heatmap_mini(dna_lst):\n",
    "    '''This function takes the output of parse_from_txt and counts codons. Several conditions are in place:\n",
    "    First, we determine how many mutations there are per gene. We discard wildtype reads (or those with zero \n",
    "    mutations. Anything with one mutation is counted. For reads with two or more mutations, any 1-bp mutations\n",
    "    are discarded, while 2+ bp mutations are counted'''\n",
    "\n",
    "    ms2_gene = ('ATGGCTTCTAACTTTACTCAGTTCGTTCTCGTCGACAATGGCGGAAC'+\n",
    "    'TGGCGACGTGACTGTCGCCCCAAGCAACTTCGCTAACGGGGTCGCTGAATGGATCAGC'+\n",
    "    'TCTAACCCGCGTTCACAGGCTTACAAAGTAACCTGTAGCGTTCGTCAGAGCTCTGCGC'+\n",
    "    'AGAATCGCAAATACACCATCAAAGTCGAGGTGCCTAAAGTGGCAACCCAGACTGTTGG'+\n",
    "    'TGGTGTAGAGCTTCCTGTAGCCGCATGGCGTTCGTACTTAAATATGGAACTAACCATT'+\n",
    "    'CCAATTTTCGCTACGAATTCCGACTGCGAGCTTATTGTTAAGGCGATGCAAGGACTCC'+\n",
    "    'TAAAAGATGGAAACCCGATTCCCTCAGCAATCGCAGCAAACTCCGGCATCTACTAA')\n",
    "    \n",
    "    intensities = np.zeros((131, 64))\n",
    "\n",
    "    codon_to_num = {'TTT': 0, 'TTG': 1, 'TTA': 2, 'TTC': 3, \n",
    "                     'CTT': 4, 'CTG': 5, 'CTA': 6, 'CTC': 7,\n",
    "                     'ATT': 8, 'ATG': 9, 'ATA': 10, 'ATC': 11,\n",
    "                     'GTT': 12, 'GTG': 13, 'GTA': 14, 'GTC': 15,\n",
    "                     'TAT': 16, 'TAG': 17, 'TAA': 18, 'TAC': 19,\n",
    "                     'CAT': 20, 'CAG': 21, 'CAA': 22, 'CAC': 23,\n",
    "                     'AAT': 24, 'AAG': 25, 'AAA': 26, 'AAC': 27,\n",
    "                     'GAT': 28, 'GAG': 29, 'GAA': 30, 'GAC': 31,\n",
    "                     'TCT': 32, 'TCG': 33, 'TCA': 34, 'TCC': 35,\n",
    "                     'CCT': 36, 'CCG': 37, 'CCA': 38, 'CCC': 39,\n",
    "                     'ACT': 40, 'ACG': 41, 'ACA': 42, 'ACC': 43,\n",
    "                     'GCT': 44, 'GCG': 45, 'GCA': 46, 'GCC': 47,\n",
    "                     'TGT': 48, 'TGG': 49, 'TGA': 50, 'TGC': 51,\n",
    "                     'CGT': 52, 'CGG': 53, 'CGA': 54, 'CGC': 55,\n",
    "                     'AGT': 56, 'AGG': 57, 'AGA': 58, 'AGC': 59,\n",
    "                     'GGT': 60, 'GGG': 61, 'GGA': 62, 'GGC' :63} \n",
    "    \n",
    "    '''Each amino acid has a number associated with it, which is also its position in AAs.\n",
    "    i.e. A = 0; S = 1; etc. AA_code allows us to move back and forth between codon and AA #.\n",
    "    Lists checked on 6/27'''\n",
    "\n",
    "    AA_code = {44: 0, 45: 0, 46: 0, 47: 0, 32: 1, 33: 1, 34: 1, 35: 1, 56: 1, 59: 1, \n",
    "                 40: 2, 41: 2, 42: 2, 43: 2, 12: 3, 13: 3, 14: 3, 15: 3, 48: 4, 51: 4, \n",
    "                 29: 5, 30: 5, 28: 6, 31: 6, 25: 7, 26: 7, 52: 8, 53: 8, 54: 8, 55: 8, \n",
    "                 57: 8, 58: 8, 21: 9, 22: 9, 24: 10, 27: 10, 9: 11, 8: 12, 10: 12, 11: 12, \n",
    "                 1: 13, 2: 13, 4: 13, 5: 13, 6: 13, 7: 13, 20: 14, 23: 14, 0: 15, 3: 15, \n",
    "                 16: 16, 19: 16, 49: 17, 60: 18, 61: 18, 62: 18, 63: 18, 36: 19, 37: 19, \n",
    "                 38: 19, 39: 19, 17: 20, 18: 20, 50: 20}\n",
    "    \n",
    "    AAs = ['A', 'S', 'T', 'V', 'C', 'E', 'D', 'K', 'R', 'Q', 'N', 'M', 'I', 'L', \n",
    "           'H', 'F', 'Y', 'W', 'G', 'P', '*']\n",
    "    \n",
    "    for dna_read in dna_lst: #dna_read counts where you are in sequencing reads\n",
    "        '''first we count how many mutations there are in a read'''\n",
    "        number_of_mutations = 0 #number_of_mutations counts number of mutations in a dna_read (seq reads)\n",
    "        for codon_number in range(131): #codon_number is the residue number in MS2 CP\n",
    "            dna_base_num = codon_number * 3 #dna_base_num is dna base pair number\n",
    "            wildtype = ms2_gene[dna_base_num:dna_base_num+3]\n",
    "            experiment_codon = dna_read[dna_base_num:dna_base_num+3]\n",
    "            if wildtype != experiment_codon:\n",
    "                #counter increases if the codon is not wildtype\n",
    "                number_of_mutations += 1\n",
    "        if number_of_mutations > 1:\n",
    "            '''if number_of_mutations > 1, get rid of any mutations that are 1 bp different from wildtype'''\n",
    "            for second_loop_codon_number in range(131): #loop through read again\n",
    "                dna_base_num = second_loop_codon_number * 3 #get DNA base pair number\n",
    "                wildtype = ms2_gene[dna_base_num:dna_base_num+3] #slice out wildtype codon\n",
    "                experiment_codon = dna_read[dna_base_num:dna_base_num+3] #slice out experiment codon\n",
    "                read_level_mutation_counter = 0 #counter for the number of basepairs that are different from wildtype\n",
    "                if wildtype != experiment_codon: #if the two codons aren't the same, count # of differences\n",
    "                    for base in range(3):\n",
    "                        if experiment_codon[base] != wildtype[base]:\n",
    "                            read_level_mutation_counter += 1\n",
    "                    if read_level_mutation_counter == 1:\n",
    "                        pass #do not count the codon if it is a 1-bp mutation\n",
    "                    elif experiment_codon in codon_to_num: #otherwise, add it to the intesities array\n",
    "                        intensities[second_loop_codon_number, codon_to_num[experiment_codon]] += 1\n",
    "        if number_of_mutations == 0:\n",
    "            '''if wildtype, toss'''\n",
    "        if number_of_mutations == 1:\n",
    "            '''keep the read if there is one total mutation'''\n",
    "            for third_loop_codon_number in range(131):\n",
    "                dna_base_num = third_loop_codon_number * 3\n",
    "                wildtype = ms2_gene[dna_base_num:dna_base_num+3]\n",
    "                experiment_codon = dna_read[dna_base_num:dna_base_num+3]\n",
    "                if wildtype != experiment_codon: #find the mutation in the read\n",
    "                    intensities[third_loop_codon_number, codon_to_num[experiment_codon]] += 1\n",
    "    return intensities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Data processing for Apparent Fitness Landscape'''\n",
    "\n",
    "def translate_NNK_array(nparray1):\n",
    "    '''Takes the codon-counted np array and turns it into a amino acid np array, discarding \n",
    "    anything that ends in C or A and is thus not from an NNK library'''\n",
    "    \n",
    "    AA_code = {44: 0, 45: 0, 46: 0, 47: 0, 32: 1, 33: 1, 34: 1, 35: 1, 56: 1, 59: 1, \n",
    "                 40: 2, 41: 2, 42: 2, 43: 2, 12: 3, 13: 3, 14: 3, 15: 3, 48: 4, 51: 4, \n",
    "                 29: 5, 30: 5, 28: 6, 31: 6, 25: 7, 26: 7, 52: 8, 53: 8, 54: 8, 55: 8, \n",
    "                 57: 8, 58: 8, 21: 9, 22: 9, 24: 10, 27: 10, 9: 11, 8: 12, 10: 12, 11: 12, \n",
    "                 1: 13, 2: 13, 4: 13, 5: 13, 6: 13, 7: 13, 20: 14, 23: 14, 0: 15, 3: 15, \n",
    "                 16: 16, 19: 16, 49: 17, 60: 18, 61: 18, 62: 18, 63: 18, 36: 19, 37: 19, \n",
    "                 38: 19, 39: 19, 17: 20, 18: 20, 50: 20}\n",
    "    \n",
    "    '''Rearranged so output is in a logical order'''\n",
    "    \n",
    "    AAs = ['A', 'S', 'T', 'V', 'C', 'E', 'D', 'K', 'R', 'Q', 'N', 'M', 'I', 'L', \n",
    "           'H', 'F', 'Y', 'W', 'G', 'P', '*']\n",
    "\n",
    "    \n",
    "    NNK_only = [0, 1, 4, 5, 8, 9, 12, 13, 16, 17, 20, 21,\n",
    "                    24, 25, 28, 29, 32, 33, 36, 37, 40, 41,\n",
    "                    44, 45, 48, 49, 52, 53, 56, 57, 60, 61]\n",
    "    \n",
    "    array_translate = np.zeros((131, 21))\n",
    "    \n",
    "    for i in NNK_only: #Loop through NNK_only (which is codon numbers excluding those ending in C or A)\n",
    "        x = AA_code[i] #Get the amino acid location where the codon row should be added\n",
    "        array_translate[:,x] += nparray1[:,i] #Add the entire row to the appropriate amino acid row\n",
    "\n",
    "    return array_translate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Data processing for Apparent Fitness Landscape'''\n",
    "\n",
    "def heatmap(array):\n",
    "    '''Combines all analyses into one function that outputs the counted array'''\n",
    "    array_p = parse_from_txt(array)\n",
    "    array_c = count_heatmap(array_p)\n",
    "    array_trans = translate_NNK_array(array_c)\n",
    "    return array_trans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Data processing for Apparent Fitness Landscape'''\n",
    "\n",
    "def heatmap_mini(array):\n",
    "    '''Combines all analyses into one function that outputs the counted array'''\n",
    "    array_p = parse_from_txt_mini(array)\n",
    "    array_c = count_heatmap_mini(array_p)\n",
    "    array_trans = translate_NNK_array(array_c)\n",
    "    return array_trans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Data processing for Apparent Fitness Landscape'''\n",
    "\n",
    "def chunk_heatmap(array):\n",
    "    '''Divides heatmap into five chunks (which is how they were cloned)\n",
    "    We compare percent abundance within each chunk between starting (plasmid) library\n",
    "    and final (VLP) library. This comparison becomes our Apparent Fitness Score'''\n",
    "    cutoffs = [0, 27, 53, 79, 105, 130]\n",
    "    sums = []\n",
    "    chunked = np.zeros((131,21))\n",
    "    for chunk in range(5):\n",
    "        summed_chunk = sum(sum(array[cutoffs[chunk]:cutoffs[chunk+1],:])) #grab the array from 0:27, 27:53, etc. Sum all points.\n",
    "        sums.append(summed_chunk) #Append that sum to a list\n",
    "    for chunk in range(5):\n",
    "        for codon in range(131):\n",
    "            for amino_acid in range(21):\n",
    "                if codon in range(cutoffs[chunk],cutoffs[chunk+1]):\n",
    "                    chunked[codon,amino_acid] = array[codon,amino_acid] / sums[chunk]\n",
    "    return chunked"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Data processing for Apparent Fitness Landscape'''\n",
    "\n",
    "def heatmap_diff(array1, array2):\n",
    "    '''Compares two heatmaps. Array 1 / Array 2, with zeros in array2 masked with np.nans'''\n",
    "    diff = np.zeros((131,21))\n",
    "    for codon in range(131):\n",
    "        for amino_acid in range(21):\n",
    "            if array2[codon,amino_acid] == 0: #if there's a zero in array2, the position becomes nan\n",
    "                diff[codon,amino_acid] = np.nan\n",
    "            else:\n",
    "                diff[codon,amino_acid] = array1[codon,amino_acid] / array2[codon,amino_acid] #else, divide, add to diff\n",
    "                \n",
    "    return diff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Data processing for Apparent Fitness Landscape'''\n",
    "\n",
    "def heatmap_avgs(array1, array2, array3):\n",
    "    '''Takes average of three heatmaps'''\n",
    "    log = np.zeros((131,21))\n",
    "    m = np.array([array1, array2, array3]) #this array of arrays means we can use axis=0 to get the mean\n",
    "    mean = np.nanmean(m, axis=0) #nanmean masks ignores nan values\n",
    "    for i in range(131):\n",
    "        for n in range(21):\n",
    "            mean[i,n] = max(mean[i,n], .0001) #if the mean is zero, replace with .0001 so the resulting log value is -4\n",
    "            log[i,n] = np.log10(mean[i,n]) #take the log of the mean value\n",
    "    return log #this is the finalized heatmap with apparent fitness scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Data processing for Apparent Fitness Landscape'''\n",
    "\n",
    "def plot_heatmap(heatmap_np, name):\n",
    "    heatmap_np = heatmap_np[1:130,:]\n",
    "\n",
    "    ytix = []\n",
    "    for i in range(1,13):\n",
    "        i = i * 10\n",
    "        ytix.append(i)\n",
    "\n",
    "\n",
    "    data = heatmap_np\n",
    "    data = np.ma.masked_invalid(data)\n",
    "    labels = 'ASTVCEDKRQNMILHFYWGP*'\n",
    "    \n",
    "    ms2_protein = ('MASNFTQFVLVDNGGTGDVTVAPSNFANGVAEWISPNPRSQAYKVTC'+\n",
    "    'SVRQSSAQNRKYTIKVEVPKVATQTVGGVELPVAAWRSYLNMELTIPIFATNSDCELIVKA'+\n",
    "    'MQGLLKDGNPIPSAIAANSGIY*')\n",
    "\n",
    "    AAs_dict = {'A': 0, 'S': 1, 'T': 2, 'V': 3, 'C': 4, 'E': 5, 'D': 6, 'K': 7, 'R': 8, 'Q': 9, \n",
    "                'N': 10, 'M': 11, 'I': 12, 'L': 13, 'H': 14, 'F': 15, 'Y': 16, 'W': 17, 'G': 18, 'P': 19, '*': 20}\n",
    "\n",
    "    wt_np = np.zeros((131,21))\n",
    "    for i in range(len(ms2_protein)):\n",
    "        wt_np[i,AAs_dict[ms2_protein[i]]] = True\n",
    "    wt_np = wt_np[1:130,:]\n",
    "\n",
    "    fig, ax = plt.subplots(figsize = (15,20))\n",
    "    \n",
    "    ax.patch.set(hatch='', color=\"silver\")\n",
    "    im = ax.pcolor(data, cmap='RdBu', vmax = 2, vmin = -2, edgecolor='black', linestyle=':', lw=1)\n",
    "    plt.colorbar(im, aspect=40)\n",
    "    plt.title('Apparent Fitness Landscape', fontsize=20)\n",
    "    plt.ylabel(\"Residue Number\", fontsize=18)\n",
    "    plt.xlabel(\"Amino Acid\", fontsize=18)\n",
    "\n",
    "    # Shift ticks to be at 0.5, 1.5, etc\n",
    "    ax.xaxis.set(ticks=np.arange(0.5, len(labels)), ticklabels=labels)\n",
    "    ax.yaxis.set(ticks=np.arange(9.5, 129, 10), ticklabels=ytix)\n",
    "    plt.tick_params(axis='y',\n",
    "                    labelsize=14)\n",
    "    \n",
    "    mask = wt_np > 0\n",
    "    for j, i in np.column_stack(np.where(mask)):\n",
    "        ax.add_patch(\n",
    "            mpatches.Rectangle(\n",
    "                (i, j),     # (x,y)\n",
    "                1,          # width\n",
    "                1,          # height\n",
    "                fill=False, \n",
    "#                 edgecolor='blue',\n",
    "                snap=False,\n",
    "                    hatch='///' # the more slashes, the denser the hash lines \n",
    "          ))\n",
    "    \n",
    "    plt.savefig(name, dpi=1000, format='pdf')\n",
    "    plt.show()\n",
    "    \n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Figure 2: The Apparent Fitness Landscape'''\n",
    "\n",
    "def heatmap_diff_bioreps(array1, array2, array3, array4, array5, array6, name):\n",
    "    '''All analysis to generate a heatmap plot'''\n",
    "    a = heatmap(array1)\n",
    "    b = heatmap(array2)\n",
    "    c = heatmap(array3)\n",
    "    d = heatmap(array4)\n",
    "    e = heatmap(array5)\n",
    "    f = heatmap(array6)\n",
    "        \n",
    "    ch_a = chunk_heatmap(a)\n",
    "    ch_b = chunk_heatmap(b)\n",
    "    ch_c = chunk_heatmap(c)\n",
    "    ch_d = chunk_heatmap(d)\n",
    "    ch_e = chunk_heatmap(e)\n",
    "    ch_f = chunk_heatmap(f)\n",
    "    \n",
    "    d_a = heatmap_diff(ch_d, ch_a)\n",
    "    e_b = heatmap_diff(ch_e, ch_b)\n",
    "    f_c = heatmap_diff(ch_f, ch_c)\n",
    "    \n",
    "    final = heatmap_avgs(d_a, e_b, f_c)\n",
    "    \n",
    "    plot_heatmap(final, name)\n",
    "        \n",
    "    return final\n",
    "\n",
    "htmp_avg = heatmap_diff_bioreps('wt-plas-r1.txt', 'wt-plas-r2.txt', 'wt-plas-r3.txt', \n",
    "                                 'wt-vlp-r1.txt', 'wt-vlp-r2.txt', 'wt-vlp-r3.txt', 'Wildtype AFL Challenge')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Figure 2: The Apparent Fitness Landscape'''\n",
    "\n",
    "def heatmap_diff_bioreps_mini(array1, array2, array3, array4, array5, array6, name):\n",
    "    '''All analysis to generate a heatmap plot'''\n",
    "    a = heatmap_mini(array1)\n",
    "    b = heatmap_mini(array2)\n",
    "    c = heatmap_mini(array3)\n",
    "    d = heatmap_mini(array4)\n",
    "    e = heatmap_mini(array5)\n",
    "    f = heatmap_mini(array6)\n",
    "        \n",
    "    ch_a = chunk_heatmap(a)\n",
    "    ch_b = chunk_heatmap(b)\n",
    "    ch_c = chunk_heatmap(c)\n",
    "    ch_d = chunk_heatmap(d)\n",
    "    ch_e = chunk_heatmap(e)\n",
    "    ch_f = chunk_heatmap(f)\n",
    "    \n",
    "    d_a = heatmap_diff(ch_d, ch_a)\n",
    "    e_b = heatmap_diff(ch_e, ch_b)\n",
    "    f_c = heatmap_diff(ch_f, ch_c)\n",
    "    \n",
    "    final = heatmap_avgs(d_a, e_b, f_c)\n",
    "    \n",
    "    plot_heatmap(final, name)\n",
    "        \n",
    "    return final\n",
    "\n",
    "htmp_avg_mini = heatmap_diff_bioreps_mini('mini-plas-r1.txt', 'mini-plas-r2.txt', 'mini-plas-r3.txt', \n",
    "                                 'mini-vlp-r1.txt', 'mini-vlp-r2.txt', 'mini-vlp-r3.txt', 'S37P AFL')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#write AFL to csv\n",
    "def using_multiindex(A, columns):\n",
    "    shape = A.shape\n",
    "    index = pd.MultiIndex.from_product([range(s)for s in shape], names=columns)\n",
    "    df = pd.DataFrame({'AFS':A.flatten()}, index=index).reset_index()\n",
    "    return df\n",
    "columns = ['Position', 'Residue']\n",
    "\n",
    "wt_df = using_multiindex(htmp_avg, columns)\n",
    "wt_df.to_csv(\"wt_AFL.csv\")\n",
    "\n",
    "wt_df = using_multiindex(htmp_avg_mini, columns)\n",
    "wt_df.to_csv(\"mini_AFL.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def shannon_calc(array):\n",
    "    '''Calculates Shannon Entropy, or a measure of diversity, on the translated heatmap'''\n",
    "    summed = np.zeros((131,1))\n",
    "    array_sum = np.zeros((131,1))\n",
    "    \n",
    "    for i in range(131):\n",
    "        summed[i,0] = sum(array[i,:]) #Sum by each row (residue number)\n",
    "    \n",
    "    array_avg = array / summed \n",
    "    #Divide each point in the array by the average of its row to get a percent abundance per residue\n",
    "    \n",
    "    for i in range(131):\n",
    "        for n in range(21):\n",
    "            array_avg[i,n] = max(array_avg[i,n], .00001) \n",
    "            #Array_avg is the maximum between .00001 and its value (exclude zeros)\n",
    "\n",
    "    array_log = array_avg * np.log10(array_avg) #Probability times the log of the probability\n",
    "    \n",
    "    for i in range(131):\n",
    "        array_sum[i,0] = -sum(array_log[i,:]) #Sum up each row by residue number\n",
    "\n",
    "    return array_sum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def shannon_avg(array1, array2, array3):\n",
    "    '''Takes three counted arrays and returns average shannon entropy'''\n",
    "    avg = np.zeros((131,1))\n",
    "    for i in range(131):\n",
    "        avg[i,0] = (array1[i,0] + array2[i,0] + array3[i,0])/3\n",
    "    return avg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def shannon_avg_from_txt(txt1, txt2, txt3):\n",
    "    \n",
    "    '''From heatmap to graphs, this function performs and plots \n",
    "    Shannon entropy calculations on three biological replicates'''\n",
    "        \n",
    "    a = heatmap(txt1)\n",
    "    b = heatmap(txt2)\n",
    "    c = heatmap(txt3)\n",
    "    \n",
    "    sh_a = shannon_calc(a)\n",
    "    sh_b = shannon_calc(b)\n",
    "    sh_c = shannon_calc(c)\n",
    "    \n",
    "    avg = shannon_avg(sh_a, sh_b, sh_c)\n",
    "    return avg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def shannon_avg_from_txt_mini(txt1, txt2, txt3):\n",
    "    \n",
    "    '''From heatmap to graphs, this function performs and plots \n",
    "    Shannon entropy calculations on three biological replicates'''\n",
    "        \n",
    "    a = heatmap_mini(txt1)\n",
    "    b = heatmap_mini(txt2)\n",
    "    c = heatmap_mini(txt3)\n",
    "    \n",
    "    sh_a = shannon_calc(a)\n",
    "    sh_b = shannon_calc(b)\n",
    "    sh_c = shannon_calc(c)\n",
    "    \n",
    "    avg = shannon_avg(sh_a, sh_b, sh_c)\n",
    "    return avg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def shannon_diff(VLP, PLAS, name):\n",
    "    '''inputs two shannon entropy arrays, returns the subtracted difference between them'''\n",
    "    shan_dif = np.subtract(VLP,PLAS)\n",
    "    \n",
    "    return shan_dif"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_shannon_heat_mult(array1, array2, array3, name):\n",
    "    \n",
    "    '''Inputs any 3 measures of shannon entropy and plots them as a heatmap'''\n",
    "\n",
    "    fig, ax = plt.subplots(figsize = (6, 8))\n",
    "    \n",
    "    plt.subplot(1,3,1)\n",
    "    im = plt.pcolor(array1, cmap='RdBu', vmax = .5, vmin = -1, linestyle=':', lw=1)\n",
    "    fig.colorbar(im)\n",
    "    plt.title('Wildtype', fontsize=14)\n",
    "    plt.ylabel(\"Residue Number\", fontsize=14)\n",
    "    plt.tick_params(\n",
    "                    axis='x',          # changes apply to the x-axis\n",
    "                    which='both',      # both major and minor ticks are affected\n",
    "                    bottom='off',      # ticks along the bottom edge are off\n",
    "                    top='off',         # ticks along the top edge are off\n",
    "                    labelbottom='off'\n",
    "                    )\n",
    "    plt.tick_params(axis='y',\n",
    "                    labelsize=14)\n",
    "    \n",
    "    plt.subplot(1,3,2)\n",
    "    im = plt.pcolor(array2, cmap='RdBu', vmax = .5, vmin = -1, linestyle=':', lw=1)\n",
    "    fig.colorbar(im)\n",
    "    plt.title('S37P', fontsize=14)\n",
    "    plt.tick_params(\n",
    "                    axis='x',          # changes apply to the x-axis\n",
    "                    which='both',      # both major and minor ticks are affected\n",
    "                    bottom='off',      # ticks along the bottom edge are off\n",
    "                    top='off',         # ticks along the top edge are off\n",
    "                    labelbottom='off'\n",
    "                    )\n",
    "    plt.tick_params(axis='y',\n",
    "                    labelsize=14)\n",
    "    \n",
    "    plt.subplot(1,3,3)\n",
    "    im = plt.pcolor(array3, cmap='RdBu', vmax = .5, vmin = -.5, linestyle=':', lw=1)\n",
    "    fig.colorbar(im)\n",
    "    plt.title('Difference', fontsize=14)\n",
    "    plt.tick_params(\n",
    "                    axis='x',          # changes apply to the x-axis\n",
    "                    which='both',      # both major and minor ticks are affected\n",
    "                    bottom='off',      # ticks along the bottom edge are off\n",
    "                    top='off',         # ticks along the top edge are off\n",
    "                    labelbottom='off'\n",
    "                    )\n",
    "    plt.tick_params(axis='y',\n",
    "                    labelsize=14)\n",
    "    \n",
    "    \n",
    "    fig.tight_layout()\n",
    "    plt.savefig(name, dpi=600, format='svg')\n",
    "    plt.show()\n",
    "    return\n",
    "\n",
    "PLAS = shannon_avg_from_txt('wt-plas-r1.txt', 'wt-plas-r2.txt', 'wt-plas-r3.txt')\n",
    "VLP = shannon_avg_from_txt('wt-vlp-r1.txt', 'wt-vlp-r2.txt', 'wt-vlp-r3.txt')\n",
    "rel_vlp_heat = shannon_diff(VLP, PLAS, 'WT_SE_Diff')\n",
    "\n",
    "PLAS_mini = shannon_avg_from_txt_mini('mini-plas-r1.txt', 'mini-plas-r2.txt', 'mini-plas-r3.txt')\n",
    "VLP_mini = shannon_avg_from_txt_mini('mini-vlp-r1.txt', 'mini-vlp-r2.txt', 'mini-vlp-r3.txt')\n",
    "rel_vlp_mini = shannon_diff(VLP_mini, PLAS_mini, 'S37P_SE_Diff')\n",
    "\n",
    "rel_vlps = shannon_diff(rel_vlp_mini, rel_vlp_heat, 'SE_diff_hm')\n",
    "plot_shannon_heat_mult(rel_vlp_heat, rel_vlp_mini, rel_vlps, \"WTvS37P_SE.svg\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
